mpi4py: run the spawn and dynamic process tests

jsquyres opened the pull request:
They're currently disabled by default upstream in mpi4py because Open MPI is failing these tests.  Explicitly enable them here in Open MPI's github action for mpi4py so that we have to fix them.

See https://github.com/mpi4py/mpi4py/pull/479 for some details.

FYI @bosilca @dalcinl @hppritcha 

Once this merges, we'll bring this over to v5.0.x.

dalcinl commented:
> Once this merges, we'll bring this over to v5.0.x.

@jsquyres Are you sure? Isn't better to leave the release branch as it now, such that any mpi4py failures would signal an new, unknown regression in the release branch?

jsquyres commented:
I thought about this overnight.  I changed up this PR a bit, and split the Github Action into 4 parts:

1. build: do everything to build, configure, and install Open MPI and mpi4py
2. run: run all the mpi4py tests with its defaults.  As of March 2024, this disables the spawn and dynamic tests, which means that the entire block of tests should pass.
3. run_spawn: run all the mpi4py tests, including the spawn tests.  As of March 2024, we know some of these tests are failing.
4. run_dynamic: run all the mpi4py tests, including the dynamic tests.  As of March 2024, we know some of these tests are failing.

Keeping the spawn and dynamics tests separate are useful because those are different failures.  So let's run them separately.

This gives us a clear green checkmark if all the mpi4py tests -- excluding spawn and dynamic tests -- pass.  Then it separately runs the spawn and dynamic tests, which, we currently expect to fail.

Looks like I don't quite have the syntax correct yet for this in the github action -- let me go fool with this over in my fork and not spam everyone here with a bunch of force pushes until I get this right.

dalcinl commented:
> Looks like I don't quite have the syntax correct yet for this in the github action 

Feel free to ping me about any doubts you have about GHA, I've been using the thing for quite a while.

jsquyres commented:
@bosilca @hppritcha I revamped the github action quite a bit since you approved it.  Please re-review.

Here's what the output looks like now in the CI on the PR (I couldn't make it also show the ✅ for the "mpi4py / build" job in the same small window):

![Screenshot 2024-03-21 at 3 58 11 PM](https://github.com/open-mpi/ompi/assets/617316/cacb2116-233d-4612-97ea-673a282b0401)

And here's what it looks like if you click through to the github action result itself:

![Screenshot 2024-03-21 at 3 59 50 PM](https://github.com/open-mpi/ompi/assets/617316/7574b2d4-3030-4fd6-8378-038c7608ca0c)

dalcinl commented on .github/workflows/ompi_mpi4py.yaml:
```diff
@@ -84,11 +91,58 @@ jobs:
+        # aren't installed by the wheel.
+        path: |
+          /opt/openmpi
+          /home/runner/.openmpi
+          /home/runner/.prte
```
No, I think GITHUB_WORKSPACE is not the right path, you truly want `$HOME`, which is the place where the `.openmpi` and `.prte` config directories are located.

Can you try doing it this way as per the [documentation](https://github.com/actions/upload-artifact/?tab=readme-ov-file#environment-variables-and-tilde-expansion) :

```suggestion
          ~/.openmpi
          ~/.prte
```

jsquyres replied:
That seemed to do the trick; thanks @dalcinl

jsquyres commented:
General consensus on the call today:

1. Do not merge this PR as-is; there is (legit) fear of normalizing CI failures.
2. Instead, make the running of the spawn/dynamic tests contingent upon the presence of a github label or special token in a the PR body or comment or something.

That way, the tests are still there and able to be run, but a developer has to do something to run them.

dalcinl commented on .github/workflows/ompi_mpi4py.yaml:
```diff
@@ -118,3 +167,135 @@ jobs:
+
+  #==============================================
+
+  run_spawn:
```
```suggestion
  run_spawn:
    if: ${{ github.event.label.name == 'mpi4py-spawn' }}
```

jsquyres replied:
This is perfect; thanks!

jsquyres replied:
It looks like `github.event.label` is the event fired when you set a label -- it doesn't check whether a label is set on the PR itself.  This Github Action doesn't fire on the `event` label.  I'll investigate more in a bit...

jsquyres replied:
@dalcinl Thanks for all that; that was most helpful in understanding what to do.

I found the proper `if` conditional:

```
if: ${{ contains(github.event.pull_request.labels.*.name, 'mpi4py-all') }}
```

This looks through all the labels on the PR right now -- not just for the labeling event -- and sees if that label is there.  A side effect here is that you have to set the label and then force the github action to run again (e.g., via a meaningless commit / push), but I'm ok with that.

dalcinl replied:
> (e.g., via a meaningless commit / push), but I'm ok with that.

You should be able to manually re-run a workflow using the web UI, look at the top right. I'm not sure if this would actually re-evaluate label changes, but worth a try...

![image](https://github.com/open-mpi/ompi/assets/294981/7832a69b-b29f-428a-95f2-a9770517c84d)

jsquyres replied:
I did try this over the weekend; it seems to send the same github context blob to the new run that it sent to the original run, so the label status is the same.

I'm ok with the PR author needing to push a new commit to the PR to get it to trigger the spawn/dynamic tests if they didn't originally label the PR that way.

dalcinl commented on .github/workflows/ompi_mpi4py.yaml:
```diff
@@ -118,3 +167,135 @@ jobs:
+
+  #==============================================
+
+  run_dynamic:
```
```suggestion
  run_dynamic:
    if: ${{ github.event.label.name == 'mpi4py-dynamic' }}
```

dalcinl commented on .github/workflows/ompi_mpi4py.yaml:
```diff
@@ -118,3 +167,135 @@ jobs:
+      run:  python test/main.py -v
+      if:   ${{ true }}
+      env:
+        MPI4PY_TEST_SPAWN: "1"
```
@jsquyres Minor nit: You can add an `env` section at the job level. In there, you set `MPI4PY_TEST_SPAWN: "1"`. That way you do not have to add the environment variable on each individual step.

dalcinl replied:
Look at the suggestion I added above. Of course, you should do the same with the mpi4py-dynamic job.

dalcinl commented on .github/workflows/ompi_mpi4py.yaml:
```diff
@@ -118,3 +167,135 @@ jobs:
+    # failing these tests.
+    needs: [build]
+    runs-on: ubuntu-latest
+    timeout-minutes: 30
```
```suggestion
    timeout-minutes: 30
    env:
        MPI4PY_TEST_SPAWN: "1"
```

jsquyres commented:
Ok, I think I got it now.  Here's what a run looks like if the `mpi4py-all` label is not on the PR:

![Screenshot 2024-03-30 at 8 16 06 PM](https://github.com/open-mpi/ompi/assets/617316/be068790-ce98-4828-80ef-99c0241d0252)

And here's what a run looks like if the `mpi4py-all` label *is* on the PR:

![Screenshot 2024-03-30 at 8 15 40 PM](https://github.com/open-mpi/ompi/assets/617316/e221d067-fdaf-4be2-b543-990bbf384fa9)

lrbison commented on .github/workflows/ompi_mpi4py.yaml:
```diff
@@ -84,37 +91,62 @@ jobs:
+      # This parameter is required, so send a meaningless
+      # environment variable name that will not affect the tests at
+      # all (i.e., the tests will be run with default values).
+      env_name: MAKE_TODAY_AN_OMPI_DAY
```
🤣
